applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,832 @@
+	case PGPS_EVT_AVAILABLE: {
+		struct gps_agps_request req;
+
+		LOG_INF("PGPS_EVT_AVAILABLE");
+		/* read out previous NRF_MODEM_GNSS_EVT_AGPS_REQ */
+		err = read_agps_req(&req);
+		if (err) {
+			/* All assistance elements as requested */
+			err = nrf_cloud_pgps_inject(prediction, &req, NULL);
+		} else {
+			/* ephemerides assistance only */
+			err = nrf_cloud_pgps_inject(prediction, NULL, NULL);
+		}
+		if (err) {
+			LOG_ERR("Unable to send prediction to modem: %d", err);
+			break;
+		}
+		err = nrf_cloud_pgps_preemptive_updates();
+		if (err) {
+			LOG_ERR("Preemptive updates error: %d", err);
+		}
+	} break;
+	/* All P-GPS predictions are available. */
+	case PGPS_EVT_READY:
+		LOG_INF("PGPS_EVT_READY");
+		break;
+	default:
+		break;
+	}
+}
+
+static void on_gnss_evt_pvt(void)
+{
+	struct nrf_modem_gnss_pvt_data_frame pvt;
+	int err;
+
+	err = nrf_modem_gnss_read((void *)&pvt, sizeof(pvt), NRF_MODEM_GNSS_DATA_PVT);
+	if (err) {
+		LOG_ERR("Failed to read GNSS PVT data, error %d", err);
+		return;
+	}
+	for (int i = 0; i < NRF_MODEM_GNSS_MAX_SATELLITES; ++i) {
+		if (pvt.sv[i].sv) { /* SV number 0 indicates no satellite */
+			LOG_DBG("SV:%3d sig: %d c/n0:%4d el:%3d az:%3d in-fix: %d unhealthy: %d",
+				pvt.sv[i].sv, pvt.sv[i].signal, pvt.sv[i].cn0,
+				pvt.sv[i].elevation, pvt.sv[i].azimuth,
+				(pvt.sv[i].flags & NRF_MODEM_GNSS_SV_FLAG_USED_IN_FIX) ? 1 : 0,
+				(pvt.sv[i].flags & NRF_MODEM_GNSS_SV_FLAG_UNHEALTHY) ? 1 : 0);
+		}
+	}
+}
+
+static void on_gnss_evt_fix(void)
+{
+	int err;
+	struct nrf_modem_gnss_pvt_data_frame pvt;
+
+	if (ttft_start != 0) {
+		LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
+		ttft_start = 0;
+	}
+	err = nrf_modem_gnss_read((void *)&pvt, sizeof(pvt), NRF_MODEM_GNSS_DATA_PVT);
+	if (err) {
+		LOG_ERR("Failed to read GNSS PVT data, error %d", err);
+		return;
+	}
+	sprintf(rsp_buf,
+		"\r\n#XGPS: %lf,%lf,%f,%f,%f,%f,\"%04u-%02u-%02u %02u:%02u:%02u\"\r\n",
+		pvt.latitude, pvt.longitude, pvt.altitude,
+		pvt.accuracy, pvt.speed, pvt.heading,
+		pvt.datetime.year, pvt.datetime.month, pvt.datetime.day,
+		pvt.datetime.hour, pvt.datetime.minute, pvt.datetime.seconds);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	if (run_type == RUN_TYPE_PGPS) {
+		struct tm gps_time = {
+			.tm_year = pvt.datetime.year - 1900,
+			.tm_mon  = pvt.datetime.month - 1,
+			.tm_mday = pvt.datetime.day,
+			.tm_hour = pvt.datetime.hour,
+			.tm_min  = pvt.datetime.minute,
+			.tm_sec  = pvt.datetime.seconds,
+		};
+
+		/* help date_time to save SNTP transactions */
+		date_time_set(&gps_time);
+		/* help nrf_cloud_pgps as most recent known location */
+		nrf_cloud_pgps_set_location(pvt.latitude, pvt.longitude);
+	}
+}
+
+static void on_gnss_evt_agps_req(void)
+{
+	if (run_type == RUN_TYPE_AGPS) {
+		k_work_submit_to_queue(&slm_work_q, &agps_req);
+	} else if (run_type == RUN_TYPE_PGPS) {
+		/* Check whether prediction data available or not */
+		k_work_submit_to_queue(&slm_work_q, &pgps_req);
+	}
```

Comment:
If you add `RUN_TYPE_AGPS_PGPS` so they can both be used together, then you'll need to handle this here.  Either add a check for this run_type in `agps_req_wk` or add a new `agps_pgps_req_wk`.

Reply:
I am marking this resolved because this use case is not required per comment from @junqingzou.